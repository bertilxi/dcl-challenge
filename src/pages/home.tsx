import {
  Button,
  Card,
  Center,
  CenterProps,
  Footer,
  Header,
  Navbar,
  Page,
  PageProps,
} from "decentraland-ui";
import { FC, PropsWithChildren } from "react";
import { Link } from "react-router";
import { useWalletStore } from "../store/wallet";

const AppPage = Page as unknown as FC<PropsWithChildren<PageProps>>;
const AppCenter = Center as unknown as FC<PropsWithChildren<CenterProps>>;

export function Home() {
  const walletStore = useWalletStore();

  return (
    <>
      <Navbar activePage="Wallet" />
      <AppPage className="App">
        <AppCenter>
          {!walletStore.isConnected ? (
            <>
              <Button
                primary
                onClick={walletStore.onConnect}
                loading={walletStore.isConnecting}
              >
                Connect
              </Button>
              {walletStore.error ? (
                <p className="error">{walletStore.error}</p>
              ) : null}
            </>
          ) : (
            <Card>
              <Header>Wallet</Header>
              <p>
                <strong>Address:</strong>&nbsp;
                {walletStore.address.slice(0, 6) +
                  "..." +
                  walletStore.address.slice(-4)}
              </p>
              <p>
                <strong>Balance:</strong>&nbsp;
                <span>{walletStore.balance.toString()} DUMMY </span>{" "}
                <Link to="/transfer" state={{ modal: location.toString() }}>
                  <Button basic>Transfer</Button>
                </Link>
              </p>
            </Card>
          )}
        </AppCenter>
      </AppPage>
      <Footer />
    </>
  );
}
