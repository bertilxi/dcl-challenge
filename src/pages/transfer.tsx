import { Button, Field, Form, Modal, ModalNavigation } from "decentraland-ui";
import { useCallback } from "react";
import { Controller, useForm } from "react-hook-form";
import { useNavigate } from "react-router";
import { useWalletStore } from "../store/wallet";

type Inputs = {
  amount: number;
  address: string;
};

export function Transfer() {
  const navigate = useNavigate();
  const walletStore = useWalletStore();
  const {
    control,
    handleSubmit,
    formState: { errors },
  } = useForm<Inputs>();

  const handleBack = useCallback(() => {
    navigate(-1);
  }, [navigate]);

  return (
    <Modal size="tiny" open={true}>
      <ModalNavigation
        title="Transfer"
        subtitle="Send tokens to an account"
        onBack={handleBack}
        onClose={handleBack}
      />
      <Form onSubmit={handleSubmit(walletStore.transfer)}>
        <Modal.Content>
          <Controller
            name="amount"
            control={control}
            rules={{ required: "The amount is required", min: 0 }}
            render={({ field }) => (
              <Field
                {...field}
                label="Amount"
                placeholder="0"
                type="number"
                error={!!errors.amount}
                message={errors.amount?.message}
              />
            )}
          />
          <Controller
            name="address"
            control={control}
            rules={{ required: "The address is required" }}
            render={({ field }) => (
              <Field
                {...field}
                label="Address"
                placeholder="0x..."
                type="address"
                error={!!errors.address}
                message={errors.address?.message}
              />
            )}
          />

          {walletStore.error ? (
            <p className="error">{walletStore.error}</p>
          ) : null}
        </Modal.Content>
        <Modal.Actions>
          <Button primary loading={walletStore.isSending}>
            Send
          </Button>
        </Modal.Actions>
      </Form>
    </Modal>
  );
}
