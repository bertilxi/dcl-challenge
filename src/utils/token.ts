import { Contract } from "ethers/contract";
import { WindowWithEthereum } from "../types";

export const windowWithEthereum = window as unknown as WindowWithEthereum;

export const TOKEN_ADDRESS = import.meta.env.VITE_TOKEN_ADDRESS;
if (!TOKEN_ADDRESS) {
  console.error(`Missing env variable VITE_TOKEN_ADDRESS`);
}

export const TOKEN_ABI = [
  "function symbol() view returns (string)",
  "function balanceOf(address) view returns (uint)",
  "function transfer(address to, uint amount)",
];

export type DummyToken = Contract & {
  symbol: () => Promise<string>;
  balanceOf: (address: string) => Promise<bigint>;
  transfer: (to: string, amount: number) => Promise<void>;
};
