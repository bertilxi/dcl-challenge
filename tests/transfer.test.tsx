import "@testing-library/jest-dom";
import { expect, test } from "vitest";
import { Transfer } from "../src/pages/transfer";
import { render } from "./utils";

test("Transfer page didn't change", async () => {
  const { asFragment } = render(<Transfer />);

  expect(asFragment()).toMatchSnapshot();
});
