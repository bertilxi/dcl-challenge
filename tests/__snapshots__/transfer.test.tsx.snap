// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transfer page didn't change 1`] = `<DocumentFragment />`;
